src/lib/keccak.ts

```diff
@@ -0,0 +1,217 @@
+import { Field } from './field.js';
+import { Gadgets } from './gadgets/gadgets.js';
+
+// KECCAK CONSTANTS
+
+// Length of the square matrix side of Keccak states
+const KECCAK_DIM = 5;
+
+// Value `l` in Keccak, ranges from 0 to 6 and determines the lane width
+const KECCAK_ELL = 6;
+
+// Width of a lane of the state, meaning the length of each word in bits (64)
+const KECCAK_WORD = 2 ** KECCAK_ELL;
+
+// Number of bytes that fit in a word (8)
+const BYTES_PER_WORD = KECCAK_WORD / 8;
+
+// Length of the state in bits, meaning the 5x5 matrix of words in bits (1600)
+const KECCAK_STATE_LENGTH = KECCAK_DIM ** 2 * KECCAK_WORD;
+
+// Number of rounds of the Keccak permutation function depending on the value `l` (24)
+const KECCAK_ROUNDS = 12 + 2 * KECCAK_ELL;
+
+// Creates the 5x5 table of rotation offset for Keccak modulo 64
+//  | x \ y |  0 |  1 |  2 |  3 |  4 |
+//  | ----- | -- | -- | -- | -- | -- |
+//  | 0     |  0 | 36 |  3 | 41 | 18 |
+//  | 1     |  1 | 44 | 10 | 45 |  2 |
+//  | 2     | 62 |  6 | 43 | 15 | 61 |
+//  | 3     | 28 | 55 | 25 | 21 | 56 |
+//  | 4     | 27 | 20 | 39 |  8 | 14 |
+const ROT_TABLE = [
+  [0, 36, 3, 41, 18],
+  [1, 44, 10, 45, 2],
+  [62, 6, 43, 15, 61],
+  [28, 55, 25, 21, 56],
+  [27, 20, 39, 8, 14],
+];
+
+// Round constants for Keccak
+// From https://keccak.team/files/Keccak-reference-3.0.pdf
+const ROUND_CONSTANTS = [
+  0x0000000000000001n,
+  0x0000000000008082n,
+  0x800000000000808an,
+  0x8000000080008000n,
+  0x000000000000808bn,
+  0x0000000080000001n,
+  0x8000000080008081n,
+  0x8000000000008009n,
+  0x000000000000008an,
+  0x0000000000000088n,
+  0x0000000080008009n,
+  0x000000008000000an,
+  0x000000008000808bn,
+  0x800000000000008bn,
+  0x8000000000008089n,
+  0x8000000000008003n,
+  0x8000000000008002n,
+  0x8000000000000080n,
+  0x000000000000800an,
+  0x800000008000000an,
+  0x8000000080008081n,
+  0x8000000000008080n,
+  0x0000000080000001n,
+  0x8000000080008008n,
+].map(Field.from);
+
+// Return a keccak state where all lanes are equal to 0
+const getKeccakStateZeros = (): Field[][] =>
+  Array.from(Array(KECCAK_DIM), (_) => Array(KECCAK_DIM).fill(Field.from(0)));
+
+// KECCAK HASH FUNCTION
+
+// ROUND TRANSFORMATION
+
+// First algrithm in the compression step of Keccak for 64-bit words.
+// C[x] = A[x,0] xor A[x,1] xor A[x,2] xor A[x,3] xor A[x,4]
+// D[x] = C[x-1] xor ROT(C[x+1], 1)
+// E[x,y] = A[x,y] xor D[x]
+// In the Keccak reference, it corresponds to the `theta` algorithm.
+// We use the first index of the state array as the x coordinate and the second index as the y coordinate.
+const theta = (state: Field[][]): Field[][] => {
+  const stateA = state;
+
+  // XOR the elements of each row together
+  // for all x in {0..4}: C[x] = A[x,0] xor A[x,1] xor A[x,2] xor A[x,3] xor A[x,4]
+  const stateC = stateA.map((row) =>
+    row.reduce((acc, next) => Gadgets.xor(acc, next, KECCAK_WORD))
```

Comment:
What's the initial value of `acc`? All zeros I suppose?

Reply:
Ah! It's just the first index of the array`row[0]`.